### PR TITLE
Add money_format function

### PR DESCRIPTION
**Pull request by kangkaisen**
For https://github.com/apache/incubator-doris/issues/1063

This PR ready for review, but ready not for merge.  

@imay  @chaoyli  please review this PR firstly, If I work on a right way, I will support decimal_v2 later.

### CODE REVIEW

File: be/src/exprs/string_functions.h [hunk 1]
```diff
@@ -137,8 +141,37 @@ class StringFunctions {
...
+    static doris_udf::StringVal money_format(doris_udf::FunctionContext* context,
+                                                 const doris_udf::DecimalVal& v);
+
+    static doris_udf::StringVal money_format(doris_udf::FunctionContext* context,
+                                                 const doris_udf::BigIntVal& v);
+
+    static doris_udf::StringVal money_format(doris_udf::FunctionContext* context,
+                                                 const doris_udf::LargeIntVal& v);
+
+    struct comma_moneypunct : std::moneypunct<char> {
```
Comment (imay) [inline be/src/exprs/string_functions.h]: class name should be camel-case
```suggestion
    struct CommaMoneypunct : std::moneypunct<char> {
```

File: be/test/exprs/string_functions_test.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (imay) [inline be/test/exprs/string_functions_test.cpp]: you should add run this test to run-ut.sh file to make regression test can run this test

File: docs/documentation/cn/sql-reference/sql-functions/string-functions/money_format.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# money_format
+
+## Syntax
+
+money_format(Number)
```
Comment (imay) [inline docs/documentation/cn/sql-reference/sql-functions/string-functions/money_format.md]:
```suggestion
VARCHAR money_format(Number)
```

File: be/src/exprs/string_functions.cpp [hunk 4]
```diff
@@ -707,4 +708,45 @@ StringVal StringFunctions::parse_url_key(
...
+}
+
+StringVal StringFunctions::money_format(FunctionContext *context, const LargeIntVal &v) {
+    if (v.is_null) {
+        return StringVal::null();
+    }
+
+    std::stringstream ss;
+    ss << v.val;
+    std::string cent_money = ss.str() + std::string("00");
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]:
```suggestion
    ss << v.val << ".00";
    std::string cent_money = ss.str();
```

**Review by imay — CHANGES_REQUESTED** [review]
You patch is in the right way, but need some changes

Thanks for your patch

File: be/src/exprs/string_functions.h [hunk 5]
```diff
@@ -137,8 +141,37 @@ class StringFunctions {
...
+        pattern do_pos_format() const override { return {{none, sign, none, value}}; }
+        pattern do_neg_format() const override { return {{none, sign, none, value}}; }
+        int do_frac_digits() const override { return 2; }
+        char_type do_thousands_sep() const override { return ','; }
+        string_type do_grouping() const override { return "\003"; }
+        string_type do_negative_sign() const override { return "-"; }
+    };
+
+    static StringVal do_money_format(FunctionContext *context, const std::string& v) {
+        std::locale comma_locale(std::locale(), new comma_moneypunct());
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/exprs/string_functions.h]: OK

File: be/src/exprs/string_functions.h (hunk 1 above)
Comment (kangkaisen) [reply be/src/exprs/string_functions.h]: OK

File: be/test/exprs/string_functions_test.cpp (hunk 2 above)
Comment (kangkaisen) [reply be/test/exprs/string_functions_test.cpp]: OK

File: docs/documentation/cn/sql-reference/sql-functions/string-functions/money_format.md (hunk 3 above)
Comment (kangkaisen) [reply docs/documentation/cn/sql-reference/sql-functions/string-functions/money_format.md]: OK

File: be/src/exprs/string_functions.cpp (hunk 4 above)
Comment (kangkaisen) [reply be/src/exprs/string_functions.cpp]: OK

**Comment by kangkaisen** [comment]
> You patch is in the right way, but need some changes
> 
> Thanks for your patch

Thanks for your review, I will update this PR.

**Comment by kangkaisen** [comment]
Update this PR and support DecimalV2.

Because the env issue, I don't run the DecimalV2 UT in StringFunctionsTest.

File: be/test/exprs/string_functions_test.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+#include "util/logging.h"
+#include "exprs/anyval_util.h"
+#include <iostream>
+#include <string>
+
+#include <gtest/gtest.h>
+
+namespace doris {
+
+    class StringFunctionsTest : public testing::Test {
```
Comment (imay) [inline be/test/exprs/string_functions_test.cpp]: there is no indent here. It is different with Java

**Comment by imay** [comment]
@kangkaisen 

Our regression test failed because of your unit test. You can check it and fix it.

image

File: be/test/exprs/string_functions_test.cpp (hunk 6 above)
Comment (kangkaisen) [reply be/test/exprs/string_functions_test.cpp]: OK

**Comment by kangkaisen** [comment]
> @kangkaisen
> 
> Our regression test failed because of your unit test. You can check it and fix it.
> 
> image
OK

**Comment by kangkaisen** [comment]
The reason for core dump  is this change:
```
-        std::locale comma_locale(std::locale(), new comma_moneypunct());
+        CommaMoneypunct comma_moneypunct;
+       std::locale comma_locale(std::locale(), &comma_moneypunct);
```
we shouldn't use stack variable, because  `std::locale`  will free  `CommaMoneypunct`. refer to
https://en.cppreference.com/w/cpp/locale/locale/locale


> Notes
Overload 7 is typically called with its second argument, f, obtained directly from a new-expression: the locale is responsible for calling the matching delete from its own destructor.

**Review by imay — APPROVED** [review]

**Pull request merged**